pkg/yang/identity.go

```diff
@@ -0,0 +1,178 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package yang
+
+import (
+	"fmt"
+	"sync"
+)
+
+// This file implements data structures and functions that relate to the
+// identity type.
+
+// identityDictionary stores a global set of identities that have been resolved
+// to be identified by their module and name.
+type identityDictionary struct {
+	mu   sync.Mutex
+	dict map[string]resolvedIdentity
+}
+
+// Global dictionary of resolved identities
+var identities = identityDictionary{dict: map[string]resolvedIdentity{}}
+
+// resolvedIdentity is an Identity that has been disambiguated.
+type resolvedIdentity struct {
+	Module   *Module
+	Identity *Identity
+}
+
+// isEmpty determines whether the resolvedIdentity struct value was defined.
+func (r resolvedIdentity) isEmpty() bool {
+	if r.Module == nil && r.Identity == nil {
+		return true
+	}
+	return false
+}
+
+// newResolvedIdentity creates a resolved identity from an identity and its
+// associated value.
+func newResolvedIdentity(m *Module, i *Identity) (string, *resolvedIdentity) {
+	r := &resolvedIdentity{
+		Module:   m,
+		Identity: i,
+	}
+	return i.PrefixedName(), r
+}
+
+func appendIfNotIn(ids []*Identity, chk *Identity) []*Identity {
+	for _, id := range ids {
+		if id == chk {
+			return ids
+		}
+	}
+	return append(ids, chk)
+}
+
+// addChildren recursively adds the identity r to ids.
+func addChildren(r *Identity, ids []*Identity) []*Identity {
+	ids = appendIfNotIn(ids, r)
+
+	// Iterate through the values of r.
+	for _, ch := range r.Values {
+		ids = addChildren(ch, ids)
+	}
+	return ids
+}
+
+// findIdentityBase returns the resolved identity that is corresponds to the
+// baseStr string in the context of the Module mod.
+func (mod *Module) findIdentityBase(baseStr string) (*resolvedIdentity, []error) {
+	var base resolvedIdentity
+	var ok bool
+	var errs []error
+
+	basePrefix, baseName := getPrefix(baseStr)
+	rootPrefix := mod.GetPrefix()
+
+	switch basePrefix {
+	case "", rootPrefix:
+		// This is a local identity which is defined within the current
+		// module
+		keyName := fmt.Sprintf("%s:%s", rootPrefix, baseName)
+		base, ok = identities.dict[keyName]
+		if !ok {
+			errs = append(errs, fmt.Errorf("can't resolve the local base %s as %s",
+				baseStr, keyName))
+		}
+		break
+	default:
+		// This is an identity which is defined within another module
+		extmod := FindModuleByPrefix(mod, basePrefix)
+		if extmod == nil {
+			errs = append(errs,
+				fmt.Errorf("can't find external module with prefix %s", basePrefix))
```

Comment:
I think you need a break after this otherwise line 110 will panic if extmod is nil.